gcovr/tests/test_args.py

```diff
@@ -0,0 +1,108 @@
+import sys
+import unittest
+from contextlib import contextmanager
+
+from ..__main__ import main, version_str
+
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+
+class CaptureObject:
+    def __init__(self, out, err, exception):
+        self.out = out
+        self.err = err
+        self.exception = exception
+
+
+class StringIOBuffered(StringIO):
+    def __init__(self):
+        self._data = None
+        StringIO.__init__(self)
+
+    def data(self):
+        self.close()
+        return self._data
+
+    def close(self):
+        try:
+            if self._data is None:
+                self.seek(0)
+                self._data = self.read()
+        finally:
+            StringIO.close(self)
+
+
+@contextmanager
+def ioscope(file):
+    try:
+        yield file
+    finally:
+        file.close()
+
+
+@contextmanager
+def capture(args):
+    newout = StringIOBuffered()
+    with ioscope(newout):
+        with ioscope(StringIOBuffered()) as newerr:
+            err, sys.stderr = sys.stderr, newerr
+            out, sys.stdout = sys.stdout, newout
+            try:
+                try:
+                    e = None
+                    main(args)
+                except BaseException as exception:
```

Comment:
It would usually be preferable to only catch `Exception` subclasses, as this will also catch Ctrl-C (`KeyboardInterrupt`). We do however want to intercept `SystemExit`. We later access the `SystemExit.code` field. So should `capture()` _only_ handle SystemExit, rather than all exceptions? The tests will still fail if we get an unexpected exception.

Reply:
I'll change this.